addons/social_share/controllers/main.py

```diff
@@ -0,0 +1,104 @@
+            url_sudo = request.env['snshare.url'].sudo().search([('campaign_id', '=', campaign_sudo.id), ('uuid', '=', uuid)])
+            target = request.env[campaign_sudo.model_id.model].sudo().browse(url_sudo.res_id).exists()
+
+        crawler = self._get_crawler_name(request)
+        if crawler and not url_sudo.is_shared:
+            request.env['bus.bus']._sendone(f'snshare_url_target-{uuid}', 'snshare/share_url_target', {
+                'message': campaign_sudo.thanks_message,
+                'reward_url': campaign_sudo.thanks_redirection,
+            })
+            url_sudo.is_shared = True
+
+        image_bytes = base64.b64decode(url_sudo._get_image()) if url_sudo else campaign_sudo._get_images_bytes(record=target)[0]
+        image_bytes = image_bytes if not small else scale_image(image_bytes, 0.5)
+        return request.make_response(image_bytes, [
+            ('Content-Type', ' image/png'),
+            ('Content-Length', len(image_bytes)),
+            ('Content-Disposition', content_disposition('card.png')),
+        ])
+
+    @route(['/snshare/campaign/<int:campaign_id>', '/snshare/campaign/<int:campaign_id>/<string:uuid>'], type='http', auth='public', sitemap=False, website=True)
+    def snshare_campaign_visitor(self, campaign_id, uuid=None):
+        """Route for users to preview their card and share it on their social platforms."""
+        campaign = request.env['snshare.campaign'].sudo().browse(campaign_id).exists()
+
+        if not campaign:
+            return request.not_found()
+
+        target = None
+
+        url = request.env['snshare.url']
+        if campaign.model_id:
+            if not uuid:
+                return request.not_found()
+            url = request.env['snshare.url'].sudo().search([('campaign_id', '=', campaign.id), ('uuid', '=', uuid)])
+            url.is_visited = True
+            target = request.env[campaign.model_id.model].sudo().browse(url.res_id).exists()
+            if not target:
+                return request.not_found()
+
+        return request.render('social_share.share_campaign_visitor', {
+            'image_url': campaign._get_card_url(uuid, small=True),
+            'link_shared_thanks_message': campaign.thanks_message if url.is_shared else '',
+            'link_shared_reward_url': campaign.thanks_redirection if url.is_shared else '',
+            'post_text': campaign.post_suggestion,
+            'redirect_url': campaign._get_redirect_url(uuid),
+            'share_url': campaign._get_redirect_url(uuid),
+            'target_name': target.display_name if target else '',
+            'uuid': uuid,
+        })
+
+    @route(['/snshare/redirect/<int:campaign_id>', '/snshare/redirect/<int:campaign_id>/<string:uuid>'], type='http', auth='public', sitemap=False, website=True)
+    def snshare_campaign_redirect(self, campaign_id, uuid=None):
+        """Route to redirect users to the target url, or display the opengraph embed text for web crawlers."""
```

Comment:
Not sure to understand this one completely ? Could gain some docstring / explanation ?